docs/networkdb.md

```diff
@@ -0,0 +1,57 @@
+Members of the cluster send each other ping messages from time to time, allowing the cluster to detect when a node has become unavailable.
+
+The information held by each node in NetworkDB is:
+
+- The set of nodes currently in the cluster (plus nodes that have recently left or failed).
+- For each peer node, the set of networks to which that node is connected.
+- For each of the node's currently-in-use networks, a set of named tables of key/value pairs.
+
+Updates are spread throughout the cluster through the gossip protocols, and nodes may have inconsistent views at any given time.
+They will eventually converge (quickly, if the network is operating well).
+Note that nodes only keep track of tables for networks to which they belong.
+Updates to a network's tables are only shared between nodes that are on that network.
+
+NetworkDB does not impose any structure on the tables.
+Other components in libnetwork use them for their own purposes.
+For example, there are tables for service discovery and load balancing,
+and the [overlay](overlay.md) driver uses NetworkDB to store routing information.
+
+All nodes in a libnetwork cluster join the gossip network.
+To do this, they need the IP address and port of at least one other member of the cluster.
+In the case of a SwarmKit cluster, for example, each Docker engine will use the IP addresses of the swarm managers as the initial join addresses.
+When joining the cluster, the new node will initially synchronise its cluster-wide state (known nodes and networks, but not tables) with at least one other node.
+The state will be mostly kept up-to-date by small UDP gossip messages, but each node will also perform a full TCP sync with another random node from time to time.
+
+Once part of the gossip cluster, a node will also send a `NodeEventTypeJoin` message, which is a custom message defined by NetworkDB.
+I'm not sure why it does this.
+
+While a node is active in the cluster, it can join and leave networks.
+When a node wants to join a network, it will send a `NetworkEventTypeJoin` message over the gossip network to the whole cluster.
+It will also perform a bulk-sync of the network-specific state (the tables) with every other node on the network being joined.
+This will allow it to get all the network-specific information quickly.
+The tables will mostly be kept up-to-date by UDP gossip messages between the nodes on that network, but
+each node in the network will also do a full TCP sync of the tables with another random node on the same network from time to time.
+
+When a node wishes to leave a network, it will send a `NetworkEventTypeLeave` via gossip. It will then delete the network's table data.
+When a node hears that another node is leaving a network, it deletes all table entries belonging to the leaving node.
+Deleting an entry in this case means marking it for deletion for a while (so that the deletion can propagate via gossip too).
```

Comment:
and more important to protect itself from receiving an old CREATE with lower version and accept it